### ISSUE TITLE
[HOLD for payment 2023-03-31] [$1000] ‘Copy url to clipboard’ is shown on right click of email address instead of ‘Copy email to clipboard’

### ISSUE THREAD

**Issue by kavimuru** [Expensify/App#15620]
If you haven’t already, check out our [contributing guidelines](https://github.com/Expensify/ReactNativeChat/blob/main/contributingGuides/CONTRIBUTING.md) for onboarding and email contributors@expensify.com to request to join our Slack channel!
___

## Action Performed:

1. Open a chat
2. Send a specific text message with email address ‘abdul@gmail.c’ or ‘$[test@gmail.com](mailto:test@gmail.com)’ or ‘=[test@gmail.com](mailto:test@gmail.com)’
3. Right click on the mail address

## Expected Result:
It should say ‘Copy email to clipboard’ similar to tooltip and click action. It should have consistency between hover/click/right click

## Actual Result:
It says ‘Copy url to clipboard’ instead of ‘copy email to clipboard’ while hover/click acts as a email.

## Workaround:
unknown

## Platforms:

Which of our officially supported platforms is this issue occurring on?

- [x] MacOS / Chrome / Safari

**Version Number:** 1.2.78-0
**Reproducible in staging?:** y
**Reproducible in production?:** y
**If this was caught during regression testing, add the test name, ID and link from TestRail:**
**Email or phone of affected tester (no customers):**
**Logs:** https://stackoverflow.com/c/expensify/questions/4856
**Notes/Photos/Videos:** 

https://user-images.githubusercontent.com/43996225/222575453-c127bfe2-e12b-48c0-a93a-6952d58fe5c1.mov

<img width="960" alt="Un titled" src="https://user-images.githubusercontent.com/43996225/222575383-2830048e-03f7-4a4d-95b1-13ed6fef320d.png">

**Expensify/Expensify Issue URL:**
**Issue reported by:** @abdulrahuman5196 
**Slack conversation:** https://expensify.slack.com/archives/C049HHMV9SM/p1677704472651439

[View all open jobs on GitHub](https://github.com/Expensify/App/issues?q=is%3Aopen+is%3Aissue+label%3A%22Help+Wanted%22)

<details><summary>Upwork Automation - Do Not Edit</summary>
    <ul>
        <li>Upwork Job URL: https://www.upwork.com/jobs/~01bc761d769cbcce0b</li>
        <li>Upwork Job ID: 1636217385617342464</li>
        <li>Last Price Increase: 2023-03-16</li>
    </ul>
</details>

**Comment by MelvinBot** [comment]
Triggered auto assignment to @kadiealexander (`Bug`), see https://stackoverflow.com/c/expensify/questions/14418 for more details.

**Comment by MelvinBot** [comment]
### Bug0 Triage Checklist ([Main S/O](https://stackoverflowteams.com/c/expensify/questions/14418))

- [ ] This "bug" occurs on a supported platform (ensure `Platforms` in OP are ✅)
- [ ] This bug is not a duplicate report (check [E/App issues](https://github.com/Expensify/App/issues) and [#expensify-bugs](https://expensify.slack.com/archives/C049HHMV9SM))
    - If it is, comment with a link to the original report, close the issue and add any novel details to the original issue instead
- [ ] This bug is reproducible using the reproduction steps in the OP. [S/O](https://stackoverflowteams.com/c/expensify/questions/14418#:~:text=The%20bug%20is%20reproducible)
    - If the reproduction steps are clear and you're unable to reproduce the bug, check with the reporter and QA first, then close the issue.
    - If the reproduction steps aren't clear and you determine the correct steps, please update the OP.
- [ ] This issue is filled out as thoroughly and clearly as possible
    - Pay special attention to the title, results, platforms where the bug occurs, and if the bug happens on staging/production.
- [ ] I have reviewed and subscribed to the linked Slack conversation to ensure Slack/Github stay in sync

**Comment by kadiealexander** [comment]
This definitely seems to be a bug, with `mailto:` links we should always show "copy email to clipboard". 

https://user-images.githubusercontent.com/59587260/223577971-d7891c52-3b2b-4131-b4e4-797253d9752b.mp4

**Comment by kadiealexander** [comment]
@jasperhuangg sorry for the delay on triaging this, I was OOO. Have you added the "reviewing" label because there is already a PR?

**Comment by jasperhuangg** [comment]
Hmm nope, there definitely shouldn't be a reviewing label on this.

**Comment by MelvinBot** [comment]
Job added to Upwork: https://www.upwork.com/jobs/~01bc761d769cbcce0b

**Comment by MelvinBot** [comment]
Current assignee @kadiealexander is eligible for the External assigner, not assigning anyone new.

**Comment by MelvinBot** [comment]
Triggered auto assignment to Contributor-plus team member for initial proposal review - @parasharrajat (`External`)

**Comment by MelvinBot** [comment]
Current assignee @jasperhuangg is eligible for the External assigner, not assigning anyone new.

**Comment by abdulrahuman5196** [comment]
## Proposal

This issue was found as part of another bug from here - https://github.com/Expensify/App/issues/15120#issuecomment-1450942212
I was the reporter of this issue and I have already provided a solution as part of the original bug report - https://expensify.slack.com/archives/C049HHMV9SM/p1677861629652999?thread_ts=1677704472.651439&cid=C049HHMV9SM

### Please re-state the problem that we are trying to solve in this issue.
‘Copy url to clipboard’ is shown on right click of email address instead of ‘Copy email to clipboard’ for specific email addresses

### What is the root cause of that problem?
Right click of email text is using different regex validation and expensiMark rule to form email has different regex. This results in inconsistency between determining a email during link formation and determining a email during right click action.

Reference:
Right click regex  
https://github.com/Expensify/App/blob/main/src/components/AnchorForCommentsOnly/BaseAnchorForCommentsOnly.js#L57
https://github.com/Expensify/expensify-common/blob/main/lib/str.js#L398
Email formation regex
https://github.com/Expensify/expensify-common/blob/main/lib/ExpensiMark.js#L102


### What changes do you think we should make in order to solve the problem?
This is a straightforward fix.
We should use the same regex for both email link creation and right click validation to find it is valid email or not.

We have already created a common markdown email formation regex(CONST.REG_EXP.MARKDOWN_EMAIL) which is used in the formation of email and email links. We can use the same regex to determine if a email is valid markdown email. I am not sure if we can change Str.isValidEmail() as it can be used in backend as well, so best approach would be create a separate isValidEmailMarkdown() and use the common regex to validate if the email is valid markdown formed.
We can also update CONST.REG_EXP.MARKDOWN_EMAIL with minor modifications if required.

Optional:
One more optional change we can do is, on right click we are passing props.displayName which will work in case of just email text like abdul@gmail.com  but it won't work in case of emails inside link like [text]-(abdul@gmail.com). As in this case props.displayName will be 'text'. So we have to passing the href value(mailTo:abdul@gmail.com) by stripping 'mailTo' which would just be the email address (abdul@gmail.com) to the isValidEmail function.
This is only if we want to show 'copy email to clipboard' incase of emails inside link as well. We can also create a separate issue if we want to deal with separately.

This would be similar to this fix - https://github.com/Expensify/App/issues/15621#issuecomment-1453778652

### What alternative solutions did you explore? (Optional)
None

**Comment by MelvinBot** [comment]
📣 @abdulrahuman5196 You have been assigned to this job by @jasperhuangg!
Please apply to this job in Upwork and leave a comment on the Github issue letting us know when we can expect a PR to be ready for review 🧑‍💻
Keep in mind: [Code of Conduct](https://github.com/Expensify/App/blob/main/CODE_OF_CONDUCT.md) | [Contributing](https://github.com/Expensify/App/blob/main/contributingGuides/CONTRIBUTING.md) 📖

**Comment by abdulrahuman5196** [comment]
@parasharrajat  I too agree. I still have doubts on that specific usecase whether it is a valid bug or not. But still it is not related to this bug fix actually. So, Maybe I will post a bug report in expensify-bugs so that someone from the team can verify if its a valid bug or not and take action accordingly.

**Comment by abdulrahuman5196** [comment]
Created a PR against expensify-common https://github.com/Expensify/expensify-common/pull/511.
@parasharrajat  and @jasperhuangg  could you assign yourselves as reviewers and review?

**Comment by abdulrahuman5196** [comment]
Created a PR against Expensify/App as well - https://github.com/Expensify/App/pull/16056
Once expensify-common PR https://github.com/Expensify/expensify-common/pull/511 is merged, I will update the package.json alone with the proper commit in the above PR.
@jasperhuangg  @parasharrajat

**Comment by MelvinBot** [comment]
`Reviewing` label has been removed, please complete the "BugZero Checklist".

**Comment by MelvinBot** [comment]
The solution for this issue has been :rocket: deployed to production :rocket: in version 1.2.88-2 and is now subject to a 7-day regression period :calendar:. Here is the list of pull requests that resolve this issue:

- https://github.com/Expensify/App/pull/16056


If no regressions arise, payment will be issued on 2023-03-31. :confetti_ball:

After the hold period is over and BZ checklist items are completed, please complete any of the applicable payments for this issue, and check them off once done.
- [ ] External issue reporter
- [ ] Contributor that fixed the issue
- [ ] Contributor+ that helped on the issue and/or PR

As a reminder, here are the bonuses/penalties that should be applied for any External issue:
- Merged PR within 3 business days of assignment - 50% **bonus**
- Merged PR more than 9 business days after assignment - 50% **penalty**

**Comment by MelvinBot** [comment]
**BugZero Checklist:** The PR fixing this issue has been merged! The following checklist ([instructions](https://stackoverflow.com/c/expensify/questions/14574)) will need to be completed before the issue can be closed:

- [ ] [@parasharrajat / @jasperhuangg] The PR that introduced the bug has been identified. Link to the PR:
- [ ] [@parasharrajat / @jasperhuangg] The offending PR has been commented on, pointing out the bug it caused and why, so the author and reviewers can learn from the mistake. Link to comment:
- [ ] [@parasharrajat / @jasperhuangg] A discussion in #expensify-bugs has been started about whether any other steps should be taken (e.g. updating the PR review checklist) in order to catch this type of bug sooner. Link to discussion:
- [ ] [@kadiealexander] Determine if we should create a regression test for this bug.
- [ ] [@parasharrajat] If we decide to create a regression test for the bug, please propose the [regression test](https://github.com/Expensify/App/blob/main/contributingGuides/REGRESSION_TEST_BEST_PRACTICES.md) steps to ensure the same bug will not reach production again.
- [ ] [@kadiealexander] Link the GH issue for creating/updating the regression test once above steps have been agreed upon:

**Comment by MelvinBot** [comment]
@parasharrajat, @jasperhuangg, @abdulrahuman5196, @kadiealexander Uh oh! This issue is overdue by 2 days. Don't forget to update your issues!

**Comment by jasperhuangg** [comment]
Still on HOLD for payment not overdue cc @kadiealexander

**Comment by kadiealexander** [comment]
Have sent Upwork contracts to @abdulrahuman5196 and @parasharrajat, just waiting for them to be accepted. :)

**Comment by abdulrahuman5196** [comment]
@kadiealexander  accepted the contract

**Comment by kadiealexander** [comment]
Thanks @abdulrahuman5196. I've paid $1750 to you just now. Here's the breakdown:

Job price: $1000
Speed bonus: $500
Reporting bonus: $250

Just waiting for @parasharrajat to accept the contract now.

**Comment by MelvinBot** [comment]
@parasharrajat, @jasperhuangg, @abdulrahuman5196, @kadiealexander Whoops! This issue is 2 days overdue. Let's get this updated quick!
